src/client/keycode.h

```diff
@@ -4,62 +4,80 @@
 
 #pragma once
 
-#include "exceptions.h"
 #include "irrlichttypes.h"
 #include <Keycodes.h>
 #include <IEventReceiver.h>
 #include <string>
+#include <unordered_map>
+#include <variant>
 
-class UnknownKeycode : public BaseException
-{
-public:
-	UnknownKeycode(const char *s) :
-		BaseException(s) {};
-};
-
-/* A key press, consisting of either an Irrlicht keycode
-   or an actual char */
-
+/* A key press, consisting of a scancode or a keycode */
 class KeyPress
 {
 public:
 	KeyPress() = default;
 
-	KeyPress(const char *name);
+	KeyPress(const std::string &name);
+
+	KeyPress(const irr::SEvent::SKeyInput &in);
+
+	// Get a string representation that is suitable for use in minetest.conf
+	std::string sym() const;
+
+	// Get a human-readable string representation
+	std::string name() const;
+
+	// Get the corresponding keycode or KEY_UNKNOWN if one is not available
+	irr::EKEY_CODE getKeycode() const;
 
-	KeyPress(const irr::SEvent::SKeyInput &in, bool prefer_character = false);
+	// Get the corresponding keychar or '\0' if one is not available
+	wchar_t getKeychar() const;
 
-	bool operator==(const KeyPress &o) const
+	// Get the scancode or 0 is one is not available
+	u32 getScancode() const
 	{
-		return (Char > 0 && Char == o.Char) || (valid_kcode(Key) && Key == o.Key);
+		if (auto pv = std::get_if<u32>(&scancode))
+			return *pv;
+		return 0;
 	}
 
-	const char *sym() const;
-	const char *name() const;
+	bool operator==(const KeyPress &o) const {
+		return scancode == o.scancode;
+	}
+	bool operator!=(const KeyPress &o) const {
+		return !(*this == o);
+	}
 
-protected:
-	static bool valid_kcode(irr::EKEY_CODE k)
+	// Check whether the keypress is valid
+	operator bool() const
 	{
-		return k > 0 && k < irr::KEY_KEY_CODES_COUNT;
+		return std::holds_alternative<irr::EKEY_CODE>(scancode) ?
+			Keycode::isValid(std::get<irr::EKEY_CODE>(scancode)) :
+			std::get<u32>(scancode) != 0;
 	}
 
-	irr::EKEY_CODE Key = irr::KEY_KEY_CODES_COUNT;
-	wchar_t Char = L'\0';
-	std::string m_name = "";
-};
+	static const KeyPress &getSpecialKey(const std::string &name);
 
-// Global defines for convenience
+private:
+	bool loadFromScancode(const std::string &name);
+	void loadFromKey(irr::EKEY_CODE keycode, wchar_t keychar);
+	std::string formatScancode() const;
 
-extern const KeyPress EscapeKey;
+	std::variant<u32, irr::EKEY_CODE> scancode = irr::KEY_UNKNOWN;
 
-extern const KeyPress LMBKey;
-extern const KeyPress MMBKey; // Middle Mouse Button
-extern const KeyPress RMBKey;
+	static std::unordered_map<std::string, KeyPress> specialKeyCache;
```

Comment:
You *could* move this variable to the source file to avoid including `unordered_map` here.

Reply:
83a8d2434